combinetf2/fitter.py

```diff
@@ -539,7 +539,136 @@ def global_impacts_parms(self):
 
         return impacts, impacts_grouped
 
-    def _expvar_optimized(self, fun_exp, skipBinByBinStat=False):
+    def _pd2ldbeta2(self, profile=False):
+        with tf.GradientTape(watch_accessed_variables=False) as t2:
+            t2.watch([self.ubeta])
+            with tf.GradientTape(watch_accessed_variables=False) as t1:
+                t1.watch([self.ubeta])
+                if profile:
+                    val = self._compute_loss(profile=profile)
+                else:
+                    # TODO this principle can probably be generalized to other parts of the code
+                    # to further reduce special cases
+
+                    # if not profiling, likelihood doesn't include the data contribution
+                    lc = self._compute_lc()
+                    if self.binByBinStat:
+                        _1, _2, beta = self._compute_yields_with_beta(
+                            profile=True, compute_norm=False, full=False
+                        )
+                        lbeta, _ = self._compute_lbeta(beta)
+                        val = lc + lbeta
+                    else:
+                        val = lc
+            pdldbeta = t1.gradient(val, self.ubeta)
+        if self.externalCovariance:
+            pd2ldbeta2_matrix = t2.jacobian(pdldbeta, self.ubeta)
+            pd2ldbeta2 = tf.linalg.LinearOperatorFullMatrix(
+                pd2ldbeta2_matrix, is_self_adjoint=True
+            )
+        else:
+            # pd2ldbeta2 is diagonal, so we can use gradient instead of jacobian
+            pd2ldbeta2_diag = t2.gradient(pdldbeta, self.ubeta)
+            pd2ldbeta2 = tf.linalg.LinearOperatorDiag(
+                pd2ldbeta2_diag, is_self_adjoint=True
+            )
+        return pd2ldbeta2
+
+    def _chi2(
+        self,
+        fun,
+        ndf_reduction=0,
+        profile=False,
+    ):
+
+        def compute_derivatives(dvars):
+            with tf.GradientTape(watch_accessed_variables=False) as t:
+                t.watch(dvars)
+                expected = self._compute_expected(
+                    fun,
+                    inclusive=True,
+                    profile=profile,
+                    full=False,
+                    need_observables=True,
+                )
+                observed = fun(None, self.nobs)
+                residuals = expected - observed
+
+                residuals_flat = tf.reshape(residuals, (-1,))
+            jacs = t.jacobian(
+                residuals_flat,
+                dvars,
+            )
+            return residuals, *jacs
+
+        if self.binByBinStat:
+            dvars = [self.x, self.ubeta, self.nobs]
+            residuals, pdresdx, pdresdbeta, pdresdobs = compute_derivatives(dvars)
+        else:
+            dvars = [self.x, self.nobs]
+            residuals, pdresdx, pdresdobs = compute_derivatives(dvars)
+            pdresdbeta = None
+
+        cov_dresdx = tf.matmul(self.cov, pdresdx, transpose_b=True)
```

Comment:
The calculation here is not correct.  What should be done is to compute the mc stat and systematics contribution to the covariance (using the "new" logic) and then add the data stat part using the "old" logic.

Instead what's being done is to compute the total uncertainty using the "new" logic and then add the data statistical contribution using the "old" logic.  (ie the data stat contribution is being partly double counted)